hasher-matcher-actioner/hmalib/fetching/tests/test_fetcher_and_bank_store.py

```diff
@@ -0,0 +1,310 @@
+                and update.tag is not None
+            ):
+                if update.signal_type in signal_types:
+                    result[update.signal_type][update.signal_value].add(update.tag)
+
+        return {
+            k: {ik: FakeSignalMetadata(iv) for ik, iv in v.items()}
+            for k, v in result.items()
+        }
+
+
+class FakeFetcher(Fetcher):
+    # Allows overriding store. Since the real fetcher must instantiate a store
+    # per collab, we need this override.
+    def __init__(
+        self,
+        signal_type_mapping: HMASignalTypeMapping,
+        banks_table: BanksTable,
+        store: BankCollabFetchStore,
+    ):
+        super().__init__(signal_type_mapping, banks_table)
+        self._fake_store = store
+
+    def get_store(self, collab: EditableCollaborationConfig) -> BankCollabFetchStore:
+        return self._fake_store
+
+
+@pytest.fixture
+def banks_table():
+    DynamoDBTableTestBase.mock_aws_credentials()
+    mddb = mock_dynamodb2()
+    mddb.start()
+    dynamodb = boto3.resource("dynamodb", region_name="us-east-1")
+
+    # Create the config table too, just in case a table needs it. Would be best
+    # if we could somehow make config_table a fixture too, but using mocked
+    # dynamodb would have become unwieldy.
+    dynamodb.create_table(**HMAConfigTestBase.get_table_definition())
+    HMAConfig.initialize(HMAConfigTestBase.TABLE_NAME)
+
+    table = dynamodb.create_table(**BanksTableTestBase.get_table_definition())
+    yield BanksTable(table, get_default_signal_type_mapping())
+    mddb.stop()
+
+
+@pytest.fixture
+def mock_enabled_signal_exchange_apis(monkeypatch):
+    def mock_get(*args, **kwargs):
+        return [
+            ToggleableSignalExchangeAPIConfig(
+                "fake",
+                "hmalib.fetching.tests.test_fetcher_and_bank_store.FakeAPI",
+                True,
+            )
+        ]
+
+    monkeypatch.setattr(ToggleableSignalExchangeAPIConfig, "get_all", mock_get)
+
+
+def get_fake_collab_config(import_as_bank_id: str) -> CollaborationConfigBase:
+    pytx_config = CollaborationConfigWithDefaults("Test State", "fake")  # type: ignore
+
+    return EditableCollaborationConfig(
+        name="does not matter",
+        collab_config_class="threatexchange.exchanges.collab_config.CollaborationConfigWithDefaults",
+        attributes_json_serialized=dataclass_dumps(pytx_config),
+        import_as_bank_id=import_as_bank_id,
+    )
+
+
+# For brevity when describing test cases.
+K1 = "key-one"
+K2 = "key-two"
+K3 = "key-three"
+
+
+def md5(n: int) -> str:
+    return f"{n:032x}"
+
+
+def pdq(n: int) -> str:
+    return f"{n:064x}"
+
+
+# Each test case is a 3 value tuple. For each case n,
+# tuple[0] is a human readable test name. We'll print this if tests fail.
+# tuple[1] is the update to be emitted by fetch_iter()
+# tuple[2] is the expected end state after all n tuple[1]s have been emitted.
+# tuple[3] if NotImplemented, indicates unsupported behavior.
```

Comment:
python now supports shape-like dicts if you want to use that for special options. 

What does unsupported mean in this context?

Reply:
Added a comment. 

Unsupported means a missing feature that is not catastrophic but should be addressed soon. An example is fetch_iter K1 [PDQ] followed by K1 [PDNA] will not remove the PDQ.

This is problematic in the long run, but in the current crowd of signal exchanges, this is unlikely to cause a problem. Because we'd need the same key to resolve to multiple hashes of the same content type for this problem to have consequences.